test/middleware/express.test.ts

```diff
@@ -0,0 +1,341 @@
+                        expect(next).toBeCalledTimes(1);
+                        expect(next).toBeCalledWith();
+
+                        // advance the timers by 1 second for the next request
+                        jest.advanceTimersByTime(20);
+                    }
+                });
+            });
+
+            describe('BLOCKS requests', () => {
+                test('A single request that exceeds capacity', () => {
+                    complexRequest = {
+                        // complexity should be 12 if 'first' is accounted for.
+                        // scalars: 1, droid: 1, reviews (5 * (1 Review, 1 episode))
+                        query: {
+                            query: `Query {
+                            scalars: {
+                                num
+                            }
+                            droid(id: 1) {
+                                name
+                            }
+                            reviews(episode: 'NEWHOPE', first: 5) {
+                                episode 
+                                stars
+                                commentary
+                            }
+                        `,
+                        },
+                    };
+                    const expectedResponse = {
+                        status: 429,
+                    };
+
+                    middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+                    expect(mockResponse.status).toBe(429);
+                    expect(nextFunction).not.toBeCalled();
+
+                    // FIXME: There are multiple functions to send a response
+                    // json, send html, sendStatus etc. How do we check at least one was called
+                    expect(mockResponse.send).toBeCalled();
+                });
+
+                test('Multiple queries that exceed token limit', () => {
+                    for (let i = 0; i < 5; i++) {
+                        // Send 5 queries of complexity 2. These should all succeed
+                        middleware(mockRequest as Request, mockResponse as Response, nextFunction);
+
+                        // advance the timers by 1 second for the next request
+                        jest.advanceTimersByTime(20);
+                    }
+
+                    // Send a 6th request that should be blocked.
+                    const next: NextFunction = jest.fn();
+                    middleware(mockRequest as Request, mockResponse as Response, next);
+                    expect(mockResponse.status).toBe(429);
+                    expect(next).not.toBeCalled();
+
+                    // FIXME: See above comment on sending responses
+                    expect(mockResponse.send).toBeCalled();
+                });
+            });
+        });
+
+        test('Uses User IP Address in Redis', async () => {
+            const client: RedisClientType = redis.createClient();
```

Comment:
How is this redis client getting into the middleware or accessing the data inside the middleware? I'm don't know enough about redis to understand what is happening here.

Reply:
We'll need to flush this out a bit since during middleware initialization the `node-redis` client would connect to an existing redis instance whereas during testing `redis-mock` does all operations in memory. This may just need to be dropped from the tests or we can use something like `NODE_ENV=TEST` to control how the middleware connects to redis during initialization.